Adding retry functionality when parallel is set to True

garethgreenaway opened the pull request:
### What does this PR do?
Making the retry state system feature available when parallel is set to True.

### What issues does this PR fix or reference?
Fixes: #61630 #61439

### Previous Behavior
When the retry options are set in a state and parallel is set to True, the job is rerun regardless of the original state results and the subsequent runs are not done in parallel.

### New Behavior
Retry functionality is properly honored with `parallel: True`

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [ ] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

ITJamie committed:
initial Proxy Module

ITJamie committed:
inital pass of restconf states

ITJamie committed:
adding codeauthor

ITJamie committed:
adding restconf module

ITJamie committed:
doc update

ITJamie committed:
doc update

ITJamie committed:
doc update

ITJamie committed:
add requirements chefk to states module

ITJamie committed:
doc update

ITJamie committed:
doc update

ITJamie committed:
doc update

ITJamie committed:
add changelog entry

ITJamie committed:
Update salt/proxy/restconf.py

Co-authored-by: Wayne Werner <waynejwerner@gmail.com>

ITJamie committed:
Update salt/proxy/restconf.py

Co-authored-by: Wayne Werner <waynejwerner@gmail.com>

ITJamie committed:
remove noqa statements

ITJamie committed:
docs update & opts defaults

ITJamie committed:
docs update

ITJamie committed:
error checks + https transport method

ITJamie committed:
docs update

ITJamie committed:
docs update

ITJamie committed:
doing a fix for commit check

ITJamie committed:
adding test and updating modules

ITJamie committed:
removing old code

ITJamie committed:
DRY

ITJamie committed:
docs update

ITJamie committed:
output updates

ITJamie committed:
prepping for deepdiff switch

ITJamie committed:
states testing

ITJamie committed:
more tests

ITJamie committed:
changed changelog from fixed to added

ITJamie committed:
force pre-commit checks

ITJamie committed:
new style of diff output that is readable

ITJamie committed:
updating tests and output

ITJamie committed:
updating output style to make YAML diff more readable

ITJamie committed:
add test for restconf module

ITJamie committed:
adjusting based on pr review

ITJamie committed:
adjusting based on pr review

ITJamie committed:
adjusting true/false verify based on pr review

ITJamie committed:
change uri to path and clean args list

Co-authored-by: Wayne Werner <waynejwerner@gmail.com>

ITJamie committed:
adjusting uri to path based on pr review

ITJamie committed:
adjusting capabilities path based on pr review

ITJamie committed:
Update salt/modules/restconf.py

Co-authored-by: Wayne Werner <waynejwerner@gmail.com>

ITJamie committed:
adjusting val based on pr review

illegalhex committed:
caps to confirm to RFC 8040

illegalhex committed:
WIP: refactor based on feedback

illegalhex committed:
WIP: changed results to be any kind of falsey val

ITJamie committed:
linting - pre-commit

ITJamie committed:
updating docs

ITJamie committed:
Update salt/states/restconf.py

Co-authored-by: Wayne Werner <waynejwerner@gmail.com>

ITJamie committed:
updating style

ITJamie committed:
fixes for lint and nox pass

ITJamie committed:
fixing

ITJamie committed:
blacken

ITJamie committed:
Update restconf.py

ITJamie committed:
add cli examples

ITJamie committed:
change logging type

s0undt3ch committed:
Also skip this test for now on macOS

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

ITJamie committed:
add pytest skip for ordereddict issue with legacy python

ITJamie committed:
add sys library for pytest mark skipif check

garethgreenaway committed:
Making the retry state system feature available when parallel is set to True.

ITJamie committed:
pre-commit fixes

Ch3LL commented on salt/state.py:
```diff
@@ -2016,11 +2016,74 @@ def _call_parallel_target(self, name, cdata, low):
+                        "state will be re-run in %s seconds",
+                        interval,
+                    )
+                    self.functions["test.sleep"](interval)
```
Why are we calling the `test.sleep` function here instead of just using python's time's lib with `time.sleep`?

garethgreenaway replied:
No particular reason, most of this code was copied from the existing `retry` functionality.  In this instance it probably makes more sense to use `time.sleep` as this is running inside it's own thread/process.  I'll update it.

Ch3LL reviewed (commented):
just a question

garethgreenaway committed:
swapping out using time.sleep Salt function for Python time.sleep.

garethgreenaway committed:
Merge branch 'master' into add_rety_to_parallel_state

s0undt3ch committed:
Switch to ``salt.utils.platform.spawning_platform()``

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Reconstruct client instances on spawning platforms

ITJamie committed:
Update salt/proxy/restconf.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

ITJamie committed:
Update salt/states/restconf.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

ITJamie committed:
Update tests/pytests/unit/modules/test_restconf.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

ITJamie committed:
Update tests/pytests/unit/states/test_restconf.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

ITJamie committed:
Update tests/pytests/unit/proxy/test_restconf.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

ITJamie committed:
pre-commit fix

Commit:
Merge branch 'master' into add_rety_to_parallel_state

s0undt3ch committed:
Skip, for now, problematic tests on spawning platforms

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

ITJamie committed:
update logging

s0undt3ch committed:
DeltaProxy minions do not work on spawning platforms

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

garethgreenaway committed:
Merge branch 'master' into add_rety_to_parallel_state

Ch3LL commented:
Looks like the windows test failures are related to this PR

garethgreenaway committed:
Skipping parallel tests on Windows.

garethgreenaway committed:
Merge branch 'master' into add_rety_to_parallel_state

krionbsd committed:
Update salt-bootstrap to 2022.03.15 release

Commit:
Merge branch 'master' into add_rety_to_parallel_state

s0undt3ch committed:
Support macOS defaulting to spawn as the default multiprocessing method

s0undt3ch committed:
Avoid a ``NameError`` when failing to initialize the process list

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

garethgreenaway committed:
Merge branch 'master' into add_rety_to_parallel_state

garethgreenaway committed:
Merge branch 'master' into add_rety_to_parallel_state

garethgreenaway committed:
Merge branch 'master' into add_rety_to_parallel_state

ITJamie committed:
Update restconf.py

garethgreenaway committed:
Merge branch 'master' into add_rety_to_parallel_state

ITJamie committed:
some rest apis dont follow the standard correctly (cisco) and need another header

garethgreenaway committed:
Merge branch 'master' into add_rety_to_parallel_state

ericham committed:
Update states.chef for version 16.x and 17.x Chef Infra Client output

s0undt3ch committed:
Use the newer nox platforms file which installs Py3.9

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Use Python 3.9 on macOS

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Drop requirements on Darwin for Py<3.9

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Same test run timeout as windows

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

ericham committed:
Support previous Chef version plus version 16.x and 17.x output.

waynew committed:
Remove legacy chef unit test

waynew committed:
Add unit tests for chef state

These tests cover the entire state, but are only concerned with the
ret["result"] of the outputs. It would be a good idea to test other
outputs as well - these tests could be modified to include assertions
about the other parts of ret. (Obviously they would need to be slightly
renamed. Maybe to `..._return_expected_ret`

Commit:
Merge branch 'master' into add_rety_to_parallel_state

s0undt3ch committed:
Don't shell out! Access the database directly and properly escape values.

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Fixes for the new macOS Catalina and BigSur CICD images

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Migrate ``integration.modules.test_mac_pkgutil`` to PyTest and functional tests.

Fixes #60819

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

Commit:
Merge branch 'master' into add_rety_to_parallel_state

s0undt3ch committed:
Add changelog for #57742

Fixes #57742

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Add changelog entry.

Fixes #55847

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Reconstruct the ``State`` class on spawned processes.

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Force a reference to ``salt.ext.tornado.iostream.StreamClosedError`` to avoid ``NameError`` issues

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Disable hack to force allowing forking on macOS

s0undt3ch committed:
Remove unused attribute

garethgreenaway committed:
When states are running in parallel, ensure that the total run time produced by the highstate outputter takes that into account.

garethgreenaway committed:
removing unrelated change

garethgreenaway committed:
Adding changelog.

garethgreenaway committed:
Fixing items, values, and keys functions in the data module.  Moving integration to functional tests.  Converting unit test over to pytest.

garethgreenaway committed:
adding changleog.

nicholasmhughes committed:
fixes saltstack/salt#62044 add ignore_missing to file.comment state

nicholasmhughes committed:
fixes saltstack/salt#61662 fix file.comment reports changes in test mode

nicholasmhughes committed:
add 61662.fixed changelog

nicholasmhughes committed:
add final test mode run after comment is successful

nicholasmhughes committed:
fix uncomment_regex

nicholasmhughes committed:
fix uncomment_regex

twangboy committed:
Adds the ability to get file version information on Windows

- Adds file.version that will get just the version
- Adds file.version_details to get additional information
- Adds tests
- Adds changelog

@amalaguti originally submitted a PR for this: #59770, but he has
since deleted his branch. Recognizing his contribution here.

garethgreenaway committed:
Skip parallel test on Windows.

nicholasmhughes committed:
fixes saltstack/salt#61946 sync_after_install immutabledict error

Commit:
skip pdbedit unit tests on ubuntu 2204 because md4 is not supported as a hash type

Commit:
fixes saltstack/salt#61944 fixed backslash literal bytestring

Commit:
refresh db at the beginning for destructive aptpkg functional tests

Commit:
fix failing test_pkgrepo_managed_absent and actually capture and keep signed-by information in salt/modules/aptpkg.py custom implementation of apt repository management

s0undt3ch committed:
Try not to trigger the GLIBC race condition.

See https://sourceware.org/bugzilla/show_bug.cgi?id=19329

Fixes #62071

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Add missing CLI example to ``network.fqdns``

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

MKLeb committed:
Update tests/unit/modules/test_pdbedit.py

Co-authored-by: Pedro Algarvio <pedro@algarvio.me>

garethgreenaway committed:
Merge branch 'master' into add_rety_to_parallel_state

s0undt3ch committed:
Sleep between submitting new jobs to thread pool

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Sleep before trying to resolve instead

Signed-off-by: Pedro Algarvio <palgarvio@vmware.com>

s0undt3ch committed:
Shorter sleeps

nicholasmhughes committed:
add configurable sync sleep for create

nicholasmhughes committed:
add testing around the particular sync command called

garethgreenaway committed:
Merge branch 'master' into add_rety_to_parallel_state

garethgreenaway committed:
Update state.py

Fixing lint.

garethgreenaway committed:
Running pre-commit manually.

Ch3LL commented:
is this a related failure? https://jenkins.saltproject.io/job/pr-macosx-catalina-x86_64-py3-pytest/job/PR-61631/22/testReport/

garethgreenaway committed:
swap out salt.utils.platform.is_windows for salt.utils.platform.spawning_platform.  Bump up duration for test_retry_option_success_parallel to 30 seconds to account for test running on Mac OS.
